device/cuda/CMakeLists.txt

```diff
@@ -1,53 +1,53 @@
-traccc_add_library( traccc_cuda cuda TYPE SHARED
-  # Utility definitions.
-  "include/traccc/cuda/utils/definitions.hpp"
-  "include/traccc/cuda/utils/make_prefix_sum_buff.hpp"
-  "src/utils/make_prefix_sum_buff.cu"
-  "include/traccc/cuda/utils/stream.hpp"
-  "src/utils/stream.cpp"
-  "src/utils/opaque_stream.hpp"
-  "src/utils/opaque_stream.cpp"
-  "src/utils/utils.hpp"
-  "src/utils/utils.cpp"
-  # Seed finding code.
-  "include/traccc/cuda/seeding/track_params_estimation.hpp"
-  "include/traccc/cuda/seeding/seed_finding.hpp"
-  "include/traccc/cuda/seeding/seeding_algorithm.hpp"
-  "include/traccc/cuda/seeding/spacepoint_binning.hpp"
-  # CCL code.
-  "include/traccc/cuda/cca/component_connection.hpp"
-  "src/seeding/track_params_estimation.cu"
-  "src/seeding/seed_finding.cu"
-  "src/seeding/spacepoint_binning.cu"
-  "src/seeding/seeding_algorithm.cpp"
-  "src/cca/component_connection.cu"
-  # Clusterization
-  "include/traccc/cuda/clusterization/clusterization_algorithm.hpp"
-  "src/clusterization/clusterization_algorithm.cu"
-  # Fitting
-  "include/traccc/cuda/fitting/fitting_algorithm.hpp"
-  "src/fitting/fitting_algorithm.cu")
-  
-target_compile_options( traccc_cuda
-  PRIVATE $<$<COMPILE_LANGUAGE:CUDA>:--expt-relaxed-constexpr> )
-target_link_libraries( traccc_cuda
-  PUBLIC traccc::core detray::core detray::utils vecmem::core covfie::core
-  PRIVATE CUDA::cudart traccc::device_common vecmem::cuda )
+#Set up the build of the traccc::cuda library.
+traccc_add_library(traccc_cuda cuda TYPE SHARED
+#Utility definitions.
+        "include/traccc/cuda/utils/definitions.hpp"
+        "include/traccc/cuda/utils/make_prefix_sum_buff.hpp"
+        "src/utils/make_prefix_sum_buff.cu"
+        "include/traccc/cuda/utils/stream.hpp"
+        "src/utils/stream.cpp"
+        "src/utils/opaque_stream.hpp"
+        "src/utils/opaque_stream.cpp"
+        "src/utils/utils.hpp"
+        "src/utils/utils.cpp"
+#Seed finding code.
+        "include/traccc/cuda/seeding/track_params_estimation.hpp"
+        "include/traccc/cuda/seeding/seed_finding.hpp"
+        "include/traccc/cuda/seeding/seeding_algorithm.hpp"
+        "include/traccc/cuda/seeding/spacepoint_binning.hpp"
+#CCL code.
+        "include/traccc/cuda/cca/component_connection.hpp"
+        "src/seeding/track_params_estimation.cu"
+        "src/seeding/seed_finding.cu"
+        "src/seeding/spacepoint_binning.cu"
+        "src/seeding/seeding_algorithm.cpp"
+        "src/cca/component_connection.cu"
+#Clusterization
+        "include/traccc/cuda/clusterization/clusterization_algorithm.hpp"
+        "src/clusterization/clusterization_algorithm.cu"
+#Fitting
+        "include/traccc/cuda/fitting/fitting_algorithm.hpp"
+        "src/fitting/fitting_algorithm.cu")
 
+target_compile_options(traccc_cuda PRIVATE
+        $<$<COMPILE_LANGUAGE:CUDA>:--expt-relaxed-constexpr
+        $<$<BOOL:${TRACCC_SETUP_PROFILER}>:-Xcompiler -finstrument-functions -fPIC>>)
```

Comment:
This is quite fancy, but not quite right.
  - You don't need to use a generator expression for evaluating `${TRACCC_SETUP_PROFILER}`. A simple

```cmake
if( TRACCC_SETUP_PROFILER )
   target_compile_options(...)
endif()
```

setup works just fine in this case.
  - The `-Xcompiler` flag forwards `-finstument-functions` to the host compiler, no? Then we should do something like:

```cmake
if( TRACCC_SETUP_PROFILER )
   target_compile_options( traccc_cuda PRIVATE
      $<COMPILE_LANGUAGE:CPP>-finstrument-functions>
      $<COMPILE_LANGUAGE:CUDA>:-Xcompiler -finstrument-functions>)
endif()
```
  - The `-fPIC` flag should not be specified like this. Please tell us why you needed to specify that explicitly. Since you're building a shared library here, position independent code generation should be the default from CMake...

Reply:
On the issue of my use of generator expressions, Stephen had expressed concern over huge chunks of code being repeated in the if statement. Thats why we resolved to use the generator expression instead.

Reply:
We use the -fPIC flag so that we can get the name of the called function in the profiler. But i can try compile the code without -fPIC and see if i can get the same output

Reply:
Huge chunks of code being repeated? You mean `$<$<COMPILE_LANGUAGE:CUDA>:...>` appearing twice? :confused: Multiple calls to `target_compile_options(...)` is definitely not a bad thing. Not in this context at least. This generator expression here confused me as well at first. Let's not try to be this clever. Let's just write simple, easy to understand code for this.

Reply:
Hi @krasznaa. Here is my implementation when using `if()`:

`if(TRACCC_PROFILER)
target_compile_options(traccc_cuda
        PRIVATE $<$<COMPILE_LANGUAGE:CPP>:--expt-relaxed-constexpr
        PRIVATE $<$<COMPILE_LANGUAGE:CUDA>:-Xcompiler -finstrument-functions>>)
elseif()
    target_compile_options(traccc_cuda
            PRIVATE $<$<COMPILE_LANGUAGE:CPP>:--expt-relaxed-constexpr>>)
    endif()`

What I meant by repetition of big chunks of code is that I had to use and if-else statement for the case where we need to have `target_compile_options()` for the case where we do not use the profiling code.